### ISSUE TITLE
EoMT for binary semantic segmentation

### ISSUE THREAD

**Issue by santimontiel** [tue-mps/eomt#3]
Dear authors,

First things first, congratulations on your fantastic work!

I would like to apply your models in a context of binary segmentation for small objects, and I have some questions about the method.

How should I train the method? I'm inputting as GT a list of `masks` with shape [1, H, W] and a tensor for `categories` full of zeros, since I only have one positive segmentation class. I added two extra Upscale layers so I can perform inference on native resolution, as H and W are small and I can deal with the extra computation overhead. Is this correct, or am I breaking the model? Is it better to perform inference on native resolution or just upsampling during post-processing?

How should I approach the post-processing of EoMT? Should I post-process with one or two classes? I applied `to_per_pixel_logits_semantic` to my last queries like this:

```python
output = self.model(x)
mask_logits, class_logits = output["mask_logits"], output["class_logits"]
mask_logits_last, class_logits_last = mask_logits[-1], class_logits[-1]
pixel_logits = to_per_pixel_logits_semantic(mask_logits_last, class_logits_last)
semantic_probs = torch.sigmoid(pixel_logits)
preds = (semantic_probs > threshold).float()
```

Applying this, I have empirically found that the optimal threshold is 0.7, since my pixel_logits are in a range from 0.5 to 1.

Thanks for your time and your attention!

Bests,
Santi

**Comment by tommiekerssies** [comment]
Hi Santi,

Thanks for your interest in our work!

I would not recommend modifying the number of upscaling blocks. Empirically, we found that with a patch size of 16, upscaling more than twice does not improve performance. If you’re aiming for finer predictions, it’s better to increase the input resolution.

Since you're doing binary segmentation with a single foreground class, I suggest avoiding `to_per_pixel_logits_semantic`, as it merges all predictions into a semantic map. Instead, treat this as an instance segmentation task and filter queries based on both their mask and class confidence.

Here's a simple way to compute confidence scores per query:

```python
# class_logits: [num_queries, 2] (foreground vs no-object)
# mask_logits: [num_queries, H, W]

class_probs = class_logits.softmax(dim=-1)[:, 0]  # probability of the foreground class
masks = mask_logits > 0

mask_scores = (
    mask_logits.sigmoid().flatten(1) * masks.flatten(1)
).sum(1) / (masks.flatten(1).sum(1) + 1e-6)

scores = class_probs * mask_scores
```

This gives a combined score by weighting the class confidence with the average confidence inside the predicted mask region. You can then filter the queries by applying a threshold on this score and keep the corresponding masks as output.

Let me know if anything’s unclear.

Best,  
Tommie

**Comment by heorhiikalaichev** [comment]
Hi @santimontiel !
How did you initialize model?
Could you share these lines of code?

**Comment by santimontiel** [comment]
> Hi Santi,
> 
> Thanks for your interest in our work!
> 
> I would not recommend modifying the number of upscaling blocks. Empirically, we found that with a patch size of 16, upscaling more than twice does not improve performance. If you’re aiming for finer predictions, it’s better to increase the input resolution.
> 
> Since you're doing binary segmentation with a single foreground class, I suggest avoiding `to_per_pixel_logits_semantic`, as it merges all predictions into a semantic map. Instead, treat this as an instance segmentation task and filter queries based on both their mask and class confidence.
> 
> Here's a simple way to compute confidence scores per query:
> 
> # class_logits: [num_queries, 2] (foreground vs no-object)
> # mask_logits: [num_queries, H, W]
> 
> class_probs = class_logits.softmax(dim=-1)[:, 0]  # probability of the foreground class
> masks = mask_logits > 0
> 
> mask_scores = (
>     mask_logits.sigmoid().flatten(1) * masks.flatten(1)
> ).sum(1) / (masks.flatten(1).sum(1) + 1e-6)
> 
> scores = class_probs * mask_scores
> This gives a combined score by weighting the class confidence with the average confidence inside the predicted mask region. You can then filter the queries by applying a threshold on this score and keep the corresponding masks as output.
> 
> Let me know if anything’s unclear.
> 
> Best, Tommie

Hi Tommie,

Thanks for your feedback! I successfully implemented your mask scoring approach. I get the best results by filtering masks with score > 0.8, just FYI.

P.S.: I'm still below my previous results with my current network using a Transformer-based architectures. I will email you if I get significant advances
